compiler/src/iree/compiler/InputConversion/Common/ImportMLProgram.cpp

```diff
@@ -77,9 +82,86 @@ class MLProgramGlobalOpPattern
         srcOpAttr.has_value()
             ? Optional<TypedAttr>(srcOpAttr.value().cast<TypedAttr>())
             : llvm::None;
+    const bool isMutable = srcOp.getIsMutable();
+    const SymbolTable::Visibility visibility = srcOp.getVisibility();
     auto globalOp = rewriter.replaceOpWithNewOp<IREE::Util::GlobalOp>(
         srcOp, srcOp.getName(), srcOp.getIsMutable(), newType, srcOpTypedAttr);
-    globalOp.setVisibility(srcOp.getVisibility());
+    globalOp.setVisibility(visibility);
+
+    if (visibility != SymbolTable::Visibility::Public) return success();
+
+    ModuleOp module = srcOp->getParentOfType<ModuleOp>();
+
+    // Default get & set.
+    std::string getterName, setterName;
+
+    auto verifyFormat = [](const std::string &format) {
+      StringRef s = format;
+      // Verify only single replacement of 0th index.
+      s = s.drop_until([](char c) { return c == '{'; });
+      if (s.empty() || !s.consume_front("{")) return failure();
+      if (!s.consume_front("0")) return failure();
+      if (!s.consume_front("}")) return failure();
+      s = s.drop_until([](char c) { return c == '{'; });
+      return success(s.empty());
+    };
+
+    if (auto v = module->getAttrOfType<DictionaryAttr>(
+            "ml_program.public_global_accessors")) {
+      // TODO(jpienaar): The attribute should be verified before here.
+      if (auto get = llvm::dyn_cast_if_present<StringAttr>(v.get("get"))) {
+        const std::string getFormat = get.str();
+        if (failed(verifyFormat(getFormat))) return failure();
+        getterName = llvm::formatv(getFormat.c_str(), globalOp.getSymName());
+      }
+      if (auto set = llvm::dyn_cast_if_present<StringAttr>(v.get("set"))) {
+        const std::string setFormat = set.str();
+        if (failed(verifyFormat(setFormat))) return failure();
+        setterName = llvm::formatv(setFormat.c_str(), globalOp.getSymName());
+      }
+    }
+    llvm::errs() << "Accessors : " << getterName << " " << setterName << "\n";
+
+    // Add public getter function.
+    if (!getterName.empty()) {
+      FunctionType funcType =
+          rewriter.getFunctionType(/*input=*/TypeRange{}, /*outputs=*/newType);
+      ImplicitLocOpBuilder b(globalOp.getLoc(), rewriter);
+      if (module.lookupSymbol(getterName)) {
```

Comment:
this is a linear scan lookup, right? are there cases where valid input programs would produce duplicates, or is this just for cases where the input program has a reserved name (incl `__`) already in use? I think we're ok with subsequent verifier failures if that's the case (would rather users get symbol name conflict errors instead of unconverted op failures they need to debug to get this match failure message)

Reply:
Yes, only where the input program already has a function that matches the one generated here. Which can be "avoided" with the format string being sufficiently different (for the python case the $ char being valid in MLIR while not in Python would make it unlikely for example). The frontend generating the file could even validate.